File: Segundo Periodo/Arquitetura de computadores/2023-2_arq1_guia_01/Java/AsciixHex.java
package Java;

import java.util.Scanner;

public class AsciixHex{

    public int toInt(String num){
        int len = num.length();
        int temp = 0;
        for(int i = 0; i < len; i++){
            temp += (num.charAt(i) - '0') * Math.pow(10,len -i -1);
        }
        return temp;
    }

    public String toLowercase(String upper){
        String temp = "";
        for(int i = 0; i < upper.length(); i++){
            char c = upper.charAt(i);
            if('A' <= c && c <= 'Z'){
                c += 32;
            }
            temp += c;
        }
        return temp;
    }


    public String toDecimal(String num){
        num = toLowercase(num);
        int len = num.length();
        int decimal = 0;
        for(int i = 0; i < len; i++){
            char numChar = num.charAt(i);
            if('a' <= numChar && numChar <= 'f'){
                decimal += (numChar - 'a' + 10) * Math.pow(16,len -i -1);
            }else{
                decimal += (numChar - '0') * Math.pow(16,len -1 -i);
            }
        }
        return decimal + "";
    }

    public String toHex(String num) {
        String temp = "";
        int numero = toInt(num);
        int resto = 0;
        while(numero > 0) {
            resto = numero % 16;
            numero /= 16;
            switch(resto){
                case 10:
                    temp = 'a' + temp;
                    break;
                case 11:
                    temp = 'b' + temp;
                    break;
                case 12:
                    temp = 'c' + temp;
                    break;
                case 13:
                    temp = 'd' + temp;
                    break;
                case 14:
                    temp = 'e' + temp;
                    break;
                case 15:
                    temp = 'f' + temp;
                    break;
                default:
                    temp = resto + temp;
                    break;
                }

        }
        return temp;
    }
    

    public String dectoHex(String recive) {
        String temp = "";
        String number = "";
    
        for (int i = 0; i < recive.length(); i++) {
            if (recive.charAt(i) != ' ' && recive.charAt(i) != '\n' && recive.charAt(i) != '\0') {
                temp += recive.charAt(i);
            } else {
                if (!temp.isEmpty()) {
                    number += toHex(temp) + " ";
                    temp = "";
                }
            }
        }
    
        if (!temp.isEmpty()) {
            number += toHex(temp);
        }
    
        return number;
    }
    
    

    public String asciiTodecimal(String ascii){
        String tempnumber = "";
        for(int i = 0; i <= ascii.length(); i++){
            if(i < ascii.length() && ascii.charAt(i) != '\n' && ascii.charAt(i) != '\0'){
                tempnumber += ' ';
                tempnumber += (int) ascii.charAt(i);
            }
        }
        
        return tempnumber;
    }

    public void asciiToHexa(String recive){
        String temp = "";

        temp = asciiTodecimal(recive);
        System.out.println("DECIMAL: " + temp);

        temp = dectoHex(temp);
        System.out.println("HEXADECIMAL: " + temp);
    }

    public String[] add(String [] arr,String append){
        int arrLength = arr.length;
        String arr2[] = new String[arrLength+1];
        for(int i = 0; i < arrLength; i++){
            arr2[i] = arr[i];
        }
        arr2[arrLength] = append;
        return arr2;
    }

    public void printArray(String[] arr){
        for(String x : arr){
            System.out.print(x + " ");
        }
        System.out.println();
    }

    public String[] toCovert(String recive){
        String temp = "";
        int len = recive.length();
        String[] arr = {};
        for(int i = 0; i <= len; i++){
            if(i < len && recive.charAt(i) != ' ' && recive.charAt(i) != '\n' && recive.charAt(i) != '\0'){
                temp += recive.charAt(i);
            }else{
                if(!temp.isEmpty()){
                    arr = add(arr, temp);
                    temp = "";
                }
            }
        }
        return arr;
    }

    public String tostring(String[] in){
        String out = "";
        for(int i = 0; i < in.length; i++){
            out += (char) toInt(in[i]);
        }
        return out;
    }

    public void HexaToAscii(String recive){
        String[] arr = toCovert(recive);
        for(int i = 0; i < arr.length; i++){
            arr[i] = toDecimal(arr[i]);
        }
        System.out.print("Decimal: ");
        printArray(arr);
        System.out.println("Ascii: " + tostring(arr));
    }

    public void executeLoop(){
        String recive = "";
        Scanner scanner = new Scanner(System.in);
        while(true){
            System.out.print("\n\n0 - sair do loop\n 1 - Ascii => Hexadecimal\n 2 - Hexadeciaml => Ascii\n Option: ");
            int op = scanner.nextInt();
            scanner.nextLine();
            if(op == 0){
                break;
            }else if(op == 1){
                System.out.print("String para converter para Ascii(16): ");
                recive = scanner.nextLine();
                System.out.print("Aperte enter!");
                scanner.nextLine();
                asciiToHexa(recive);
            }else if(op == 2){
                System.out.print("Ascii(16) para converter para String: ");
                recive = scanner.nextLine();
                System.out.print("Aperte enter!");
                scanner.nextLine();
                HexaToAscii(recive);
            }

            System.out.print("Aperte enter!");
            scanner.nextLine();
        }
    }
}

File: Segundo Periodo/Arquitetura de computadores/2023-2_arq1_guia_01/Python/AsciixHex.py
class AsciixHex:
    def to_int(self, num):
        len_num = len(num)
        temp = 0
        for i in range(len_num):
            temp += (int(num[i]) - 0) * (10 ** (len_num - i - 1))
        return temp

    def to_lowercase(self, upper):
        temp = ""
        for c in upper:
            if 'A' <= c <= 'Z':
                c = chr(ord(c) + 32)
            temp += c
        return temp

    def to_decimal(self, num):
        num = self.to_lowercase(num)
        len_num = len(num)
        decimal = 0
        for i in range(len_num):
            num_char = num[i]
            if 'a' <= num_char <= 'f':
                decimal += (ord(num_char) - ord('a') + 10) * (16 ** (len_num - i - 1))
            else:
                decimal += (int(num_char) - 0) * (16 ** (len_num - 1 - i))
        return str(decimal)

    def to_hex(self, num):
        temp = ""
        numero = self.to_int(num)
        resto = 0
        while numero > 0:
            resto = numero % 16
            numero //= 16
            if resto == 10:
                temp = 'a' + temp
            elif resto == 11:
                temp = 'b' + temp
            elif resto == 12:
                temp = 'c' + temp
            elif resto == 13:
                temp = 'd' + temp
            elif resto == 14:
                temp = 'e' + temp
            elif resto == 15:
                temp = 'f' + temp
            else:
                temp = str(resto) + temp
        return temp

    def dec_to_hex(self, recive):
        temp = ""
        number = ""
        for char in recive:
            if char != ' ' and char != '\n' and char != '\0':
                temp += char
            else:
                if temp:
                    number += self.to_hex(temp) + " "
                    temp = ""
        if temp:
            number += self.to_hex(temp)
        return number

    def ascii_to_decimal(self, ascii_str):
        temp_number = ""
        for char in ascii_str:
            if char != '\n' and char != '\0':
                temp_number += ' ' + str(ord(char))
        return temp_number

    def ascii_to_hexa(self, recive):
        temp = ""
        temp = self.ascii_to_decimal(recive)
        print("DECIMAL:", temp)
        temp = self.dec_to_hex(temp)
        print("HEXADECIMAL:", temp)

    def add(self, arr, append):
        arr_length = len(arr)
        arr2 = arr.copy()
        arr2.append(append)
        return arr2

    def print_array(self, arr):
        print(" ".join(arr))

    def to_covert(self, recive):
        temp = ""
        len_recive = len(recive)
        arr = []
        for i in range(len_recive):
            if recive[i] != ' ' and recive[i] != '\n' and recive[i] != '\0':
                temp += recive[i]
            else:
                if temp:
                    arr = self.add(arr, temp)
                    temp = ""
        if temp:  # Añadir este bloque para manejar el último elemento
            arr = self.add(arr, temp)
        return arr

    def to_string(self, arr):
        out = ""
        for item in arr:
            out += chr(self.to_int(item))
        return out

    def hexa_to_ascii(self, recive):
        arr = self.to_covert(recive)
        arr = [self.to_decimal(item) for item in arr]
        print("Decimal:", arr)
        print("Ascii:", self.to_string(arr))

        


